File: solutions/Leetcode_677/leetcode_677.py
from collections import defaultdict


class TrieNode:
    def __init__(self):
        self.child = defaultdict(TrieNode)
        self.val = 0


class MapSum:

    def __init__(self):
        self._root = TrieNode()
        self._dict = defaultdict(int)

    def insert(self, key: str, val: int) -> None:
        diff = val - self._dict[key]
        self._dict[key] = val
        current = self._root
        for x in key:
            current = current.child[x]
            current.val += diff

    def sum(self, prefix: str) -> int:
        current = self._root
        for x in prefix:
            if x not in current.child:
                return 0
            current = current.child[x]
        return current.val

File: solutions/Leetcode_677/leetcode_677.cpp
#include <string>
#include <unordered_map>

struct TrieNode
{
public:
    TrieNode* child[26] = {};
    int val = 0;
};

class MapSum {
public:
    TrieNode root;
    std::unordered_map<std::string, int> dict;

    MapSum() {}
    
    void insert(std::string key, int val) {
        int diff = val - dict[key];
        dict[key] = val;
        TrieNode* current = &root;
        for (char x : key) {
            x -= 'a';
            if (current->child[x] == nullptr)
                current->child[x] = new TrieNode();
            current = current->child[x];
            current->val += diff;
        }
    }
    
    int sum(std::string prefix) {
        TrieNode* current = &root;
        for (char x : prefix) {
            x -= 'a';
            if (current->child[x] == nullptr)
                return 0;
            current = current->child[x];
        }
        return current->val;
    }
};